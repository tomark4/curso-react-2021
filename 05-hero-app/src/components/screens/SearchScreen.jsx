import { useState, useEffect } from "react";
import HeroCard from "../shared/HeroCard";
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { getHeoresByName } from "../../selectors/getHeroesByName";

const SearchScreen = ({history}) => {

    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    const [data, setData] = useState([]);
    const [search, setSearch] = useState(q);

    useEffect(() => {
        setData(getHeoresByName(q));
    },[q]);

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push("?q="+search);
    }

    return (
        <div>
            <div className="row">
                <div className="col-md-5">
                    <h4>Search form</h4>
                    <form onSubmit={handleSubmit}>
                        <div className="form-group">
                            <input type="text" className="form-control"
                            name="search"
                            value={search}
                            onChange={(e) => setSearch(e.target.value)}/>
                        </div>
                        <div className="form-group">
                            <button className="btn btn-outline-primary btn-block">Buscar</button>
                        </div>
                    </form>
                </div>
                <div className="col-md-7">
                    <h4>Heroes</h4>
                    <hr/>
                    {data.length === 0 &&
                        <div className="alert alert-danger">
                            No hay data
                        </div>
                    }
                    {data.map( item => (
                        <div className="my-2" key={item.id}>
                            <HeroCard heroe={item}
                           />
                        </div>
                    ))}
                </div>
            </div>
        </div>
    )
}

export default SearchScreen
