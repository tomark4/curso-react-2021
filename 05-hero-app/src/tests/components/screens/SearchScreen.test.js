import {mount} from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import SearchScreen from '../../../components/screens/SearchScreen';

describe('pruebas en el search screen component', () => {

  test('debe de mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot();
    expect( wrapper.find('.alert-danger').text().trim()).toBe('No hay data');
  });

  test('debe de mostrar a batman y el input con el valor del query string', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper.find('input').prop('value')).toBe('batman');
  });

  test('debe de mostrar si no se encuentrar el hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=jose']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay data');
    expect(wrapper.find('.alert-danger').exists()).toBe(true);
  });

  test('debe de llamar el push de history', () => {
    const history = {
      push: jest.fn()
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search"
        component={() => <SearchScreen history={history} />}/>
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change',{
      target:{
        name: 'search',
        value:'batman'
      }
    })

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });

    expect( history.push ).toHaveBeenCalledWith('?q=batman');
  });

});
