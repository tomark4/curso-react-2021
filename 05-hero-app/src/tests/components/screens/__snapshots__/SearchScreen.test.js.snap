// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el search screen component debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="search"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <button
            className="btn btn-outline-primary btn-block"
          >
            Buscar
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h4>
        Heroes
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay data
      </div>
    </div>
  </div>
</div>
`;
